### ISSUE TITLE
CanvasKit: move image decoding to web workers

### ISSUE THREAD

**Issue by yjbanov** [flutter/flutter#63397]
See also https://bugs.chromium.org/p/skia/issues/detail?id=10360

Image decoding in Flutter is already [asynchronous](https://api.flutter.dev/flutter/dart-ui/instantiateImageCodec.html), so this would be a non-breaking change. There are two ways to move image decoding to web workers:

- Use WASM thread support and use a web-worker as a second thread. It can be used for image decoding without blocking the main UI thread. In the future we can move font decoders there too. Image data can be passed efficiently by sharing WASM memory.
- Use Canvas2D on a web-worker. Image data can be passed efficiently using transferrable arrays.

**Comment by IchordeDionysos** [comment]
@yjbanov what would be the expected reduction in bytes required for canvas kit?

**Comment by yjbanov** [comment]
@IchordeDionysos should be a few 100s of KB.

**Comment by yjbanov** [comment]
I scoped this down a bit. The main goal is to eliminate jank caused by image decoding blocking the main UI thread. Size reduction would be nice to have, but I think we'll have more flexibility if we addressed it separately.

**Comment by linxuebin1990** [comment]
@yjbanov I am very interested in this issue. This is a cool feature. Is there any progress?

**Comment by yjbanov** [comment]
I'm experimenting with `ImageDecoder` instead of web workers here: https://github.com/flutter/engine/pull/29419

**Comment by linxuebin1990** [comment]
@yjbanov This feature is very cool ！

**Comment by linxuebin1990** [comment]
@yjbanov We implemented this feature by rewriting the Canvaskit js code in our own local engine,based on the createImageBitmap and Offscreen Canvas interfaces. In our web project, the `FPS` of the image list has been increased from `30` to `54`. But we currently only support the decoding of static pictures, so we are looking forward to this feature for you.

**Comment by yjbanov** [comment]
Yep, that's one of the big advantages of WebCodecs. They support animated images. They are also lighter than web workers on memory, and they provide a faster path to the GPU.

Update on the PR: I'm currently waiting on a fix from CanvasKit. `makeImageFromTextureSource` doesn't know how to hand off an image from one WebGL context to another, so if you have a platform view, the image fails to render. Once I confirm the fix and we roll to the new build of CanvasKit, I'll merge the PR.
